lmdeploy/pytorch/disagg/backend/mooncake.py

```diff
@@ -1,28 +1,256 @@
+            # Sample output:
+            # device                 node GUID
+            # ------              ----------------
+            # erdma_0             02163efffe3fc264
+            # erdma_1             02163efffe3fc317
+            lines = result.stdout.strip().split('\n')
+            for line in lines[2:]:  # Skip header lines
+                if line.strip():
+                    device_name = line.split()[0].strip()
+                    rdma_nics.append(device_name)
+    except Exception as e:
+        print(f'Error executing ibv_devices command: {e}')
+
+    return rdma_nics
+
+
+def get_local_ip_by_remote() -> str:
+    # try ipv4
+    s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
+    try:
+        s.connect(('8.8.8.8', 80))  # Doesn't need to be reachable
+        return s.getsockname()[0]
+    except Exception:
+        pass
+
+    # try ipv6
+    try:
+        s = socket.socket(socket.AF_INET6, socket.SOCK_DGRAM)
+        # Google's public DNS server, see
+        # https://developers.google.com/speed/public-dns/docs/using#addresses
+        s.connect(('2001:4860:4860::8888', 80))  # Doesn't need to be reachable
+        return s.getsockname()[0]
+    except Exception:
+        raise ValueError('Can not get local ip')
+
+
+class MooncakeMigrationManagement:
+    """Manages migration for a single connection in Mooncake backend."""
+
+    def __init__(self, init_request: DistServeInitRequest):
+        try:
+            from mooncake.engine import TransferEngine
+        except ImportError as e:
+            raise ImportError('Please install mooncake by following the instructions at '
+                              'https://github.com/kvcache-ai/Mooncake/blob/main/doc/en/build.md '
+                              'to run LMDeploy with MooncakeBackend.') from e
+
+        self.rank = init_request.rank
+        self.local_engine_config: DistServeEngineConfig = init_request.local_engine_config
+        self.remote_engine_config: DistServeEngineConfig = init_request.remote_engine_config
+        self.local_engine_id = init_request.local_engine_id
+        self.remote_engine_id = init_request.remote_engine_id
+
+        self.engine = TransferEngine()
+        self.hostname = get_local_ip_by_remote()
+
+        # Get all RDMA information once during initialization
+        self.ibv_devices = get_rdma_nics()
+
+        self.local_kv_table: Dict[str, Dict] = {}
+        self.remote_kv_table: Dict[str, Dict] = {}
+        self.remote_url: str = ''  # Store remote URL for this connection
+
+        # Initialize the p2p connection
+        self._initialize_p2p(init_request)
+
+        self.port: int = self.engine.get_rpc_port()
+
+    def _initialize_p2p(self, init_request: DistServeInitRequest):
+        """Initialize p2p connection for this specific link."""
+        # TODO: Support more types of metadata_server
+        # e.g. "etcd://192.168.0.137:2379"
+        metadata_server = 'P2PHANDSHAKE'
```

Comment:
what is metadata_server used for?

Reply:
There are two modes: (1) 'P2PHANDSHAKE' (a magic string): no metadata server for maintaining connection information, which is intended for small-scale PD disaggregation, and (2) support for etcd/redis/http_server as the centralized server for larger-scale PD disaggregation.